MdeModulePkg/Bus/Pci/PciBusDxe/PciDoe.c

```diff
@@ -19,98 +19,209 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
-                           1,
-                           &DoeControl
-                           );
+  Status = PciIoDevice->PciIo.Pci.Read (
+                                    &PciIoDevice->PciIo,
+                                    EfiPciIoWidthUint32,
+                                    PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_CONTROL_OFFSET,
+                                    1,
+                                    &DoeControl
+                                    );
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
 
   // Set Go bit
   DoeControl.Bits.DoeGo = 1;
 
   DEBUG ((DEBUG_INFO, "Setting DOE Go bit\n"));
 
-  PciIoDevice->PciIo.Pci.Write (
-                           &PciIoDevice->PciIo,
-                           EfiPciIoWidthUint32,
-                           PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_CONTROL_OFFSET,
-                           1,
-                           &DoeControl
-                           );
+  Status = PciIoDevice->PciIo.Pci.Write (
+                                    &PciIoDevice->PciIo,
+                                    EfiPciIoWidthUint32,
+                                    PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_CONTROL_OFFSET,
+                                    1,
+                                    &DoeControl
+                                    );
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
+
+  return EFI_SUCCESS;
 }
 
-VOID
+/** Wait for the DOE Data Object Ready (DOR) bit to be set
+
+  @param[in]  PciIoDevice    A pointer to the EFI_PCI_IO_PROTOCOL instance.
+
+  @retval EFI_SUCCESS Source was updated to support Handler.
+  @retval EFI_UNSUPPORTED  Hardware could not be programmed.
+  @retval EFI_TIMEOUT  Did not complete in timeout period.
+**/
+EFI_STATUS
 DoeWaitForStatusDor (
   IN PCI_IO_DEVICE  *PciIoDevice
   )
 {
+  EFI_STATUS                  Status;
   PCI_EXPRESS_REG_DOE_STATUS  DoeStatus;
+  UINTN                       Timeout;
 
   //
   // Read DOE Status
   //
-  PciIoDevice->PciIo.Pci.Read (
-                           &PciIoDevice->PciIo,
-                           EfiPciIoWidthUint32,
-                           PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_STATUS_OFFSET,
-                           1,
-                           &DoeStatus
-                           );
+  Status = PciIoDevice->PciIo.Pci.Read (
+                                    &PciIoDevice->PciIo,
+                                    EfiPciIoWidthUint32,
+                                    PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_STATUS_OFFSET,
+                                    1,
+                                    &DoeStatus
+                                    );
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
 
   DEBUG ((DEBUG_INFO, "Waiting for DOE Data Object Read\n"));
 
+  Timeout = 10000;

```

Comment:
What are the units if this Timeout and where does this hard coded constant of 10000 come from?

Reply:
I fixed the value and clarified that we are waiting for 1 second as defined in the PCIe spec